src/app/icd/client/CheckInHandler.h

```diff
@@ -0,0 +1,94 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines objects for a CHIP CheckInMessage unsolicited
+ *      handler
+ *
+ */
+
+#pragma once
+
+#include <app/icd/client/CheckInDelegate.h>
+#include <app/icd/client/DefaultICDClientStorage.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+
+namespace chip {
+namespace app {
+class CheckInMessageHandler : public Messaging::ExchangeDelegate, public Messaging::UnsolicitedMessageHandler
+{
+
+public:
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage, CheckInDelegate * delegate);
+    void Shutdown();
+
+    virtual ~CheckInMessageHandler() = default;
+
+protected:
+    // ExchangeDelegate
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override;
+
+    // UnsolicitedMessageHandler
+    CHIP_ERROR OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate) override;
+
+    // TODO : Follow up to check if this really needs to be a pure virtual function in Exchange delegate
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
+
+    Messaging::ExchangeMessageDispatch & GetMessageDispatch() override { return CheckInExchangeDispatch::Instance(); }
+
+private:
+    class CheckInExchangeDispatch : public Messaging::ExchangeMessageDispatch
+    {
+    public:
+        static ExchangeMessageDispatch & Instance()
+        {
+            static CheckInExchangeDispatch instance;
+            return instance;
+        }
+
+        CheckInExchangeDispatch() {}
+        ~CheckInExchangeDispatch() override {}
+
+    protected:
+        bool MessagePermitted(Protocols::Id, uint8_t type) override
+        {
+            return type == to_underlying(Protocols::SecureChannel::MsgType::ICD_CheckIn);
+        }
+        bool IsEncryptionRequired() const override { return false; }
+    };
+
+    /**
+     * @brief Used by the application to set a new key to avoid counter rollover problems.
+     *
+     * @param[in] clientInfo clientInfo object
+     * @param[in] keyData New key data to use to re-register the client with the server
+     */
+    CHIP_ERROR RegisterClientWithNewKey(ICDClientInfo & clientInfo, const ByteSpan keyData);
+
+    Messaging::ExchangeManager * mpExchangeManager = nullptr;
+    CheckInDelegate * mpCheckInDelegate            = nullptr;
+    Messaging::ExchangeManager * GetExchangeManager(void) const { return mpExchangeManager; }
```

Comment:
Just for readability, i would move this function before the members of the class.